plugin-server/src/worker/ingestion/event-pipeline/cookielessServerHashStep.ts

```diff
@@ -0,0 +1,585 @@
+    host: string | undefined
+    timezone: string | undefined
+    timestampMs: number
+    teamId: number
+    hashExtra: string | undefined
+} {
+    const userAgent = event.properties?.['$raw_user_agent']
+    const ip = event.properties?.['$ip']
+    const host = event.properties?.['$host']
+    const timezone = event.properties?.['$timezone']
+    const hashExtra = event.properties?.[COOKIELESS_EXTRA_HASH_CONTENTS_PROPERTY]
+    const timestampMs = DateTime.fromISO(timestamp).toMillis()
+    const teamId = event.team_id
+
+    return { userAgent, ip, host, timezone, timestampMs, teamId, hashExtra }
+}
+
+const localSaltMap: Record<string, Uint32Array> = {}
+const mutex = new ConcurrencyController(1)
+
+export async function getSaltForDay(
+    db: DB,
+    timestamp: number,
+    eventTimeZone: string | undefined,
+    teamtimeZone: string
+): Promise<Uint32Array> {
+    // get the day based on the timezone
+    const yyyymmdd = toYYYYMMDDInTimezoneSafe(timestamp, eventTimeZone, teamtimeZone)
+
+    if (!isCalendarDateValid(yyyymmdd)) {
+        throw new Error('Date is out of range')
+    }
+
+    // see if we have it locally
+    if (localSaltMap[yyyymmdd]) {
+        return localSaltMap[yyyymmdd]
+    }
+
+    // get the salt for the day from redis, but only do this once for this node process
+    return mutex.run({
+        fn: async (): Promise<Uint32Array> => {
+            // check if we got the salt while waiting for the mutex
+            if (localSaltMap[yyyymmdd]) {
+                return localSaltMap[yyyymmdd]
+            }
+
+            // try to get it from redis instead
+            const saltBase64 = await db.redisGet<string | null>(
+                `cookieless_salt:${yyyymmdd}`,
+                null,
+                'cookielessServerHashStep'
+            )
+            if (saltBase64) {
+                const salt = base64StringToUint32Array(saltBase64)
+                localSaltMap[yyyymmdd] = salt
+                return salt
+            }
+
+            // try to write a new one to redis, but don't overwrite
+            const newSaltParts = createRandomUint32x4()
+            const setResult = await db.redisSetNX(
+                `cookieless_salt:${yyyymmdd}`,
+                uint32ArrayToBase64String(newSaltParts),
+                'cookielessServerHashStep',
+                SALT_TTL_SECONDS
+            )
+            if (setResult === 'OK') {
+                localSaltMap[yyyymmdd] = newSaltParts
+                return newSaltParts
+            }
+
+            // if we couldn't write, it means that it exists in redis already
+            const saltBase64Retry = await db.redisGet<string | null>(
+                `cookieless_salt:${yyyymmdd}`,
+                null,
+                'cookielessServerHashStep'
+            )
+            if (!saltBase64Retry) {
+                throw new Error('Failed to get salt from redis')
+            }
+            return base64StringToUint32Array(saltBase64Retry)
```

Comment:
You forgot to save it in the `localSaltMap` here

Reply:
Fixed! Though the new tests don't cover this exact case (handling this race condition)

Reply:
Fixed!